### PR TITLE
[Merged by Bors] - feat: characterize totally disconnected sets in orders

### PR DESCRIPTION
**Pull request by fpvandoorn**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Intervals/OrdConnected.lean [hunk 1]
```diff
@@ -292,6 +292,30 @@ protected theorem _root_.IsAntichain.ordConnected (hs : IsAntichain (· ≤ ·)
...
+  intros u hu v hv _huv w hw
+  have h2w : w ∈ Icc x y := ⟨hu.2.1.trans hw.1, hw.2.trans hv.2.2⟩
+  refine ⟨?_, h2w⟩
+  by_cases hwu : w = u
+  · subst w; exact hu.1
+  by_cases hwv : w = v
+  · subst w; exact hv.1
+  refine h ⟨h2w.1.lt_of_ne ?_, h2w.2.lt_of_ne ?_⟩
+  · rintro rfl; exact hwu <| hu.2.1.antisymm hw.1
+  · rintro rfl; exact hwv <| hw.2.antisymm hv.2.2
```
Comment (urkud) [inline Mathlib/Data/Set/Intervals/OrdConnected.lean:306]: :golfing:
```suggestion
lemma ordConnected_inter_Icc_of_subset (h : Ioo x y ⊆ s) : OrdConnected (s ∩ Icc x y) :=
  ordConnected_of_Ioo fun _u ⟨_, hu, _⟩ _v ⟨_, _, hv⟩ _ ↦
    (Ioo_subset_Ioo hu hv).trans <| subset_inter h (Ioo_subset_Icc_self)
```

File: Mathlib/Data/Set/Intervals/OrdConnected.lean [hunk 2]
```diff
@@ -292,6 +292,30 @@ protected theorem _root_.IsAntichain.ordConnected (hs : IsAntichain (· ≤ ·)
...
+
+lemma ordConnected_inter_Icc_iff (hx : x ∈ s) (hy : y ∈ s) :
+    OrdConnected (s ∩ Icc x y) ↔ Ioo x y ⊆ s := by
+  refine ⟨fun h z hz ↦ ?_, ordConnected_inter_Icc_of_subset⟩
+  simp_rw [ordConnected_iff] at h
+  have hxy : x ≤ y := hz.1.trans hz.2 |>.le
+  exact h x ⟨hx, le_rfl, hxy⟩ y ⟨hy, hxy, le_rfl⟩ hxy ⟨hz.1.le, hz.2.le⟩ |>.1
```
Comment (urkud) [inline Mathlib/Data/Set/Intervals/OrdConnected.lean:313]: :golfing: 
```suggestion
  refine ⟨fun h ↦ Ioo_subset_Icc_self.trans fun z hz ↦ ?_, ordConnected_inter_Icc_of_subset⟩
  have hxy : x ≤ y := hz.1.trans hz.2
  exact h.out ⟨hx, left_mem_Icc.2 hxy⟩ ⟨hy, right_mem_Icc.2 hxy⟩ hz |>.1
```

File: Mathlib/Topology/Order/IntermediateValue.lean [hunk 3]
```diff
@@ -516,6 +516,19 @@ theorem setOf_isPreconnected_eq_of_ordered :
...
+  simp_rw [IsTotallyDisconnected, isPreconnected_iff_ordConnected, ← not_nontrivial_iff]
+  refine ⟨fun h x hx y hy hxy ↦ ?_, fun h t hts ht ⟨x, hx, y, hy, hxy⟩ ↦ ?_⟩
+  · simp_rw [← mem_Ioo, ← not_ordConnected_inter_Icc_iff hx hy]
+    refine fun hs ↦ h _ (inter_subset_left _ _) hs ?_
+    exact ⟨x, ⟨hx, le_rfl, hxy.le⟩, y, ⟨hy, hxy.le, le_rfl⟩, hxy.ne⟩
+  · obtain hxy|hxy := hxy.lt_or_lt
+    · obtain ⟨z, h1z, h2z⟩ := h x (hts hx) y (hts hy) hxy
+      exact h1z <| hts <| ht.1 hx hy ⟨h2z.1.le, h2z.2.le⟩
+    · obtain ⟨z, h1z, h2z⟩ := h y (hts hy) x (hts hx) hxy
+      exact h1z <| hts <| ht.1 hy hx ⟨h2z.1.le, h2z.2.le⟩
```
Comment (urkud) [inline Mathlib/Topology/Order/IntermediateValue.lean:530]:
```suggestion
  simp_rw [IsTotallyDisconnected, isPreconnected_iff_ordConnected, ← not_nontrivial_iff,
    nontrivial_iff_exists_lt]
  refine ⟨fun h x hx y hy hxy ↦ ?_, fun h t hts ht ⟨x, hx, y, hy, hxy⟩ ↦ ?_⟩
  · simp_rw [← mem_Ioo, ← not_ordConnected_inter_Icc_iff hx hy]
    refine fun hs ↦ h _ (inter_subset_left _ _) hs ?_
    exact ⟨x, ⟨hx, le_rfl, hxy.le⟩, y, ⟨hy, hxy.le, le_rfl⟩, hxy⟩
  · obtain ⟨z, h1z, h2z⟩ := h x (hts hx) y (hts hy) hxy
    exact h1z <| hts <| ht.1 hx hy ⟨h2z.1.le, h2z.2.le⟩
```

**Comment by urkud** [comment]
I added some optional golf suggestions. Also, please add a docstring to `lemma isTotallyDisconnected_iff_lt`.

Otherwise LGTM. Thanks!

bors d+

**Comment by fpvandoorn** [comment]
Thanks for the golfs!

**Pull request closed without merging**